lib/std/time/TimeZone.zig

```diff
@@ -0,0 +1,299 @@
+        if (occur < 0) return error.Malformed; // rfc8536: occur [...] MUST be nonnegative
+        if (i > 0 and leapseconds[i - 1].occurrence + 2419199 > occur) return error.Malformed; // rfc8536: occur [...] each later value MUST be at least 2419199 greater than the previous value
+        if (occur > std.math.maxInt(i48)) return error.Malformed; // Unreasonably far into the future
+
+        const corr = try reader.readInt(i32, .big);
+        if (i == 0 and corr != -1 and corr != 1) return error.Malformed; // rfc8536: The correction value in the first leap-second record, if present, MUST be either one (1) or minus one (-1)
+        if (i > 0 and leapseconds[i - 1].correction != corr + 1 and leapseconds[i - 1].correction != corr - 1) return error.Malformed; // rfc8536: The correction values in adjacent leap-second records MUST differ by exactly one (1)
+        if (corr > std.math.maxInt(i16)) return error.Malformed; // Unreasonably large correction
+
+        leapseconds[i] = .{
+            .occurrence = @as(i48, @intCast(occur)),
+            .correction = @as(i16, @intCast(corr)),
+        };
+    }
+
+    // Parse standard/wall indicators
+    i = 0;
+    while (i < header.counts.isstdcnt) : (i += 1) {
+        const stdtime = try reader.readByte();
+        if (stdtime == 1) {
+            timetypes[i].flags |= 0x02;
+        }
+    }
+
+    // Parse UT/local indicators
+    i = 0;
+    while (i < header.counts.isutcnt) : (i += 1) {
+        const ut = try reader.readByte();
+        if (ut == 1) {
+            timetypes[i].flags |= 0x04;
+            if (!timetypes[i].standardTimeIndicator()) return error.Malformed; // rfc8536: standard/wall value MUST be one (1) if the UT/local value is one (1)
+        }
+    }
+
+    // Footer
+    var footer: ?[]u8 = null;
+    if (!legacy) {
+        if ((try reader.readByte()) != '\n') return error.Malformed; // An rfc8536 footer must start with a newline
+        var footerdata_buf: [128]u8 = undefined;
+        const footer_mem = reader.readUntilDelimiter(&footerdata_buf, '\n') catch |err| switch (err) {
+            error.StreamTooLong => return error.OverlargeFooter, // Read more than 128 bytes, much larger than any reasonable POSIX TZ string
+            else => |e| return e,
+        };
+        if (footer_mem.len != 0) {
+            footer = try allocator.dupe(u8, footer_mem);
+        }
+    }
+    errdefer if (footer) |ft| allocator.free(ft);
+
+    return TimeZone{
+        .transitions = transitions,
+        .timetypes = timetypes,
+        .leapseconds = leapseconds,
+        .footer = footer,
+    };
+}
+
+pub fn deinit(self: *TimeZone, allocator: std.mem.Allocator) void {
+    if (self.footer) |footer| {
+        allocator.free(footer);
+    }
+    allocator.free(self.leapseconds);
+    allocator.free(self.transitions);
+    allocator.free(self.timetypes);
+}
+
+/// Project UTC timestamp to this time zone.
```

Comment:
Does this cover a fall-back to LMT if the tzfile neither defines an offset for given Unix time nor a general POSIX rule? (v1 TZif)
Btw. Unix time, not UTC ;-)

Reply:
No, currently it does only the most basic lookup for the offset. I'll add the leap seconds but mainly I was hoping for someone more familiar with the format to handle the edge cases so that I don't need to learn more about it.

Reply:
Thinking this through once more, I think if you have a IANA tz db file loaded (TZif), it should specify an LMT offset as the first entry in the `timetypes` array. I'm not sure if this is historically "correct" since the concept of time zones as we have them today didn't exist a 150 years ago. But to me it seems "more" correct than extending a POSIX rule indefinitely into the past.
For the future, the story is a bit different; if the tzfile does not define any transitions after 2038, you end up with a fixed UTC offset, which doesn't seem right. But as for the past, extending a DST rule into the future indefinitely also seems... optimistic?